File: services/user-service/src/main/java/com/visor/school/userservice/controller/ParentController.java
package com.visor.school.userservice.controller;

import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.visor.school.common.api.ApiResponse;
import com.visor.school.userservice.dto.LinkStudentRequest;
import com.visor.school.userservice.dto.ParentStudentResponse;
import com.visor.school.userservice.model.Parent;
import com.visor.school.userservice.service.ParentService;

import jakarta.validation.Valid;

/**
 * Parent controller for managing parent-student relationships
 */
@RestController
@RequestMapping("/v1/parents")
public class ParentController {

    private final ParentService parentService;

    public ParentController(ParentService parentService) {
        this.parentService = parentService;
    }

    /**
     * Link a student to a parent
     */
    @PostMapping("/{userId}/students/{studentId}")
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    public ResponseEntity<ApiResponse<Map<String, String>>> linkStudentToParent(
        @PathVariable UUID userId,
        @PathVariable UUID studentId,
        @Valid @RequestBody LinkStudentRequest request
    ) {
        parentService.linkStudentToParent(
            userId,
            studentId,
            request.relationship(),
            request.isPrimary() != null ? request.isPrimary() : false
        );

        return ResponseEntity.status(HttpStatus.CREATED)
            .body(
                ApiResponse.success(
                    Map.of(
                        "message", "Student linked to parent successfully",
                        "userId", userId.toString(),
                        "studentId", studentId.toString()
                    )
                )
            );
    }

    /**
     * Get all children (students) for a parent
     */
    @GetMapping("/{parentId}/students")
    @PreAuthorize("hasRole('ADMINISTRATOR') or @securityContextService.isCurrentUserId(#parentId)")
    public ResponseEntity<ApiResponse<List<ParentStudentResponse>>> getChildren(@PathVariable UUID parentId) {
        List<Parent> parents = parentService.getChildrenByParent(parentId);

        List<ParentStudentResponse> responses = parents.stream()
            .map(ParentStudentResponse::from)
            .collect(Collectors.toList());

        return ResponseEntity.ok(ApiResponse.success(responses));
    }

    /**
     * Unlink student from parent
     */
    @DeleteMapping("/{userId}/students/{studentId}")
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    public ResponseEntity<ApiResponse<Map<String, String>>> unlinkStudentFromParent(
        @PathVariable UUID userId,
        @PathVariable UUID studentId
    ) {
        parentService.unlinkStudentFromParent(userId, studentId);

        return ResponseEntity.ok(
            ApiResponse.success(
                Map.of("message", "Student unlinked from parent successfully")
            )
        );
    }
}

File: services/user-service/src/main/kotlin/com/visor/school/userservice/controller/ParentController.kt
package com.visor.school.userservice.controller

import com.visor.school.common.api.ApiResponse
import com.visor.school.userservice.model.Relationship
import com.visor.school.userservice.service.ParentService
import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import java.util.UUID

/**
 * Parent controller for managing parent-student relationships
 */
@RestController
@RequestMapping("/v1/parents")
class ParentController(
    private val parentService: ParentService
) {

    /**
     * Link a student to a parent
     */
    @PostMapping("/{userId}/students/{studentId}")
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    fun linkStudentToParent(
        @PathVariable userId: UUID,
        @PathVariable studentId: UUID,
        @Valid @RequestBody request: LinkStudentRequest
    ): ResponseEntity<ApiResponse<Map<String, String>>> {
        parentService.linkStudentToParent(
            userId = userId,
            studentId = studentId,
            relationship = request.relationship,
            isPrimary = request.isPrimary ?: false
        )

        return ResponseEntity.status(HttpStatus.CREATED)
            .body(
                ApiResponse.success(
                    mapOf(
                        "message" to "Student linked to parent successfully",
                        "userId" to userId.toString(),
                        "studentId" to studentId.toString()
                    )
                )
            )
    }

    /**
     * Get all children (students) for a parent
     */
    @GetMapping("/{parentId}/students")
    @PreAuthorize("hasRole('ADMINISTRATOR') or @securityContextService.isCurrentUserId(#parentId)")
    fun getChildren(@PathVariable parentId: UUID): ResponseEntity<ApiResponse<List<ParentStudentResponse>>> {
        val parents = parentService.getChildrenByParent(parentId)
        
        val responses = parents.map {
            ParentStudentResponse(
                parentId = it.userId,
                studentId = it.studentId,
                relationship = it.relationship.name,
                isPrimary = it.isPrimary
            )
        }

        return ResponseEntity.ok(ApiResponse.success(responses))
    }

    /**
     * Unlink student from parent
     */
    @DeleteMapping("/{userId}/students/{studentId}")
    @PreAuthorize("hasRole('ADMINISTRATOR')")
    fun unlinkStudentFromParent(
        @PathVariable userId: UUID,
        @PathVariable studentId: UUID
    ): ResponseEntity<ApiResponse<Map<String, String>>> {
        parentService.unlinkStudentFromParent(userId, studentId)

        return ResponseEntity.ok(
            ApiResponse.success(
                mapOf("message" to "Student unlinked from parent successfully")
            )
        )
    }
}

data class LinkStudentRequest(
    @field:NotNull
    val relationship: Relationship,
    val isPrimary: Boolean? = false
)

data class ParentStudentResponse(
    val parentId: UUID,
    val studentId: UUID,
    val relationship: String,
    val isPrimary: Boolean
)

